csl/core/engine.lua

```diff
@@ -0,0 +1,967 @@
+end
+
+function CslEngine:_preprocess ()
+   -- Handle locale overrides
+   if self.locales[self.locale.lang] then -- Direct language match
+      local override = CslLocale(self.locales[self.locale.lang])
+      SU.debug("csl", "Locale override found for " .. self.locale.lang)
+      self:_merge_locales(self.locale, override)
+   else
+      for lang, locale in pairs(self.locales) do -- Fuzzy language matching
+         if self.locale.lang:sub(1, #lang) == lang then
+            local override = CslLocale(locale)
+            SU.debug("csl", "Locale override found for " .. self.locale.lang .. " -> " .. lang)
+            self:_merge_locales(self.locale, override)
+         end
+      end
+   end
+end
+
+-- GROUP LOGIC (tracking variables in groups, conditional rendering)
+
+function CslEngine:_enterGroup ()
+   self.groupState.count = self.groupState.count + 1
+   SU.debug("csl", "Enter group", self.groupState.count, "level", #self.groupQueue)
+
+   table.insert(self.groupQueue, self.groupState)
+   self.groupState = { variables = {}, count = 0 }
+end
+
+function CslEngine:_leaveGroup (rendered)
+   -- Groups implicitly act as a conditional: if all variables that are called
+   -- are empty, the group is suppressed.
+   -- But the group is kept if no variable is called.
+   local emptyVariables = true
+   local hasVariables = false
+   for _, cond in pairs(self.groupState.variables) do
+      hasVariables = true
+      if cond then -- non-empty variable found
+         emptyVariables = false
+         break
+      end
+   end
+   local suppressGroup = hasVariables and emptyVariables
+   if suppressGroup then
+      rendered = nil -- Suppress group
+   end
+   self.groupState = table.remove(self.groupQueue)
+   -- A nested non-empty group is treated as a non-empty variable for the
+   -- purposes of determining suppression of the outer group.
+   -- So add a pseudo-variable for the inner group to the outer group, to
+   -- track this.
+   if not suppressGroup then
+      local groupCond = "_group_" .. self.groupState.count
+      self:_addGroupVariable(groupCond, true)
+   end
+   SU.debug("csl", "Leave group", self.groupState.count, "level", #self.groupQueue, suppressGroup and "(suppressed)" or "(rendered)")
+   return rendered
+end
+
+function CslEngine:_addGroupVariable (variable, value)
+   SU.debug("csl", "Group variable", variable, value and "true" or "false")
+   self.groupState.variables[variable] = value and true or false
+end
+
+-- RENDERING ATTRIBUTES (strip-periods, affixes, formatting, text-case, display, quotes)
+
+function CslEngine:_render_stripPeriods (t, options)
+   if t and options["strip-periods"] and t:sub(-1) == "." then
+      t = t:sub(1, -2)
+   end
+   return t
+end
+
+function CslEngine:_render_affixes (t, options)
+   if not t then
+      return
+   end
+   if options.prefix then
+      -- replace by localized prefix
+      local pref = options.prefix:gsub("[,;:]", function (c) return self.punctuation[c] or c end)
```

Comment:
We don't really care with SILE, as we do our own automated space handling for French (so we always get it right, removing and inserting non-breaking space automatically around concerned punctuation signs).

But _in a more global picture,_ this is probably "too clever" (same remark applies to the suffix and delimiter): 
 - chicago-author-date-fr _already_ encodes the non-breakable space as `&160;` in the delimiters, and affixes, so we now end up with two of them in the output... (one comes from the fr-FR locale for "colon", "semicolon"). As noted, SILE doesn't really care, the output will be corrected, but that's weird...)
 - If using other styles not ported to French (even the default generic chicago-author-date), if we remove this delim/affix localization, we won't get the non-breakable space in the output. (But again we don't care, SILE inserts them as appropriate)
 
On the other hand, there must be a reason for CSL local to have some punctuation terms. 'Cause if we remove this delim/affix localization, the terms aren't used anywhere -- and I don't understand why they exist in the first place!

Possible solutions:
 1. Remove this delim/affix punctuation localization, it wasn't supposed to exist...
 2. Or keep it only if the style default-locale exists and matches the locale lang... though I see the troubles coming, fr-CA etc. are going to be messy then...

We could question the CSL folks regarding the purpose of the "colon", "semicolon" terms etc. vs. the intended use of delimiters and affixes, but it's likely a hole in the specification... So the best course of action could be (1), simplifying the code, esp. since it has no effect with SILE's own good French support.

Reply:
> We could question the CSL folks regarding the purpose of the "colon", "semicolon" terms etc. vs. the intended use of delimiters and affixes, but it's likely a hole in the specification..

I [asked](https://github.com/citation-style-language/locales/issues/315) -- and the status seems to be that such replacements are not part of CSL 1.0.2.

Reply:
Fixed (kept behind a non-standard extension)